File: mobile/src/main/java/net/mm2d/android/upnp/cds/chapter/SonyFetcherFactory.java
/*
 * Copyright (c) 2017 大前良介 (OHMAE Ryosuke)
 *
 * This software is released under the MIT License.
 * http://opensource.org/licenses/MIT
 */

package net.mm2d.android.upnp.cds.chapter;

import android.text.TextUtils;

import net.mm2d.android.upnp.cds.CdsObject;
import net.mm2d.upnp.HttpClient;
import net.mm2d.util.XmlUtils;

import org.w3c.dom.Element;
import org.w3c.dom.Node;
import org.xml.sax.SAXException;

import java.io.IOException;
import java.net.URL;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.concurrent.TimeUnit;

import javax.xml.parsers.ParserConfigurationException;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import io.reactivex.Single;
import io.reactivex.SingleOnSubscribe;
import io.reactivex.schedulers.Schedulers;

/**
 * @author <a href="mailto:ryo@mm2d.net">大前良介 (OHMAE Ryosuke)</a>
 */
class SonyFetcherFactory implements FetcherFactory {
    private static final String CHAPTER_INFO = "av:chapterInfo";
    private static final String ROOT_NODE = "contentInfo";
    private static final String LIST_NODE = "content_chapter_info";
    private static final String ITEM_NODE = "chapter";
    private static final String TIME_NODE = "chapter_point";

    @Nullable
    @Override
    public Single<List<Integer>> create(@NonNull final CdsObject object) {
        final String url = object.getValue(CHAPTER_INFO);
        if (TextUtils.isEmpty(url)) {
            return null;
        }
        return Single.create((SingleOnSubscribe<List<Integer>>) emitter -> {
            try {
                final String xml = new HttpClient(false).downloadString(new URL(url));
                emitter.onSuccess(parseChapterInfo(xml));
            } catch (IOException | ParserConfigurationException | SAXException ignored) {
                emitter.onSuccess(Collections.emptyList());
            }
        }).subscribeOn(Schedulers.io());
    }

    @NonNull
    private List<Integer> parseChapterInfo(@NonNull final String xml)
            throws ParserConfigurationException, SAXException, IOException {
        if (TextUtils.isEmpty(xml)) {
            return Collections.emptyList();
        }
        final Element root = XmlUtils.newDocument(false, xml).getDocumentElement();
        if (root == null || !root.getNodeName().equals(ROOT_NODE)) {
            return Collections.emptyList();
        }
        final Element content = findChildElementByNodeName(root, LIST_NODE);
        if (content == null) {
            return Collections.emptyList();
        }
        final List<Integer> result = new ArrayList<>();
        for (Node node = content.getFirstChild(); node != null; node = node.getNextSibling()) {
            if (node.getNodeType() != Node.ELEMENT_NODE
                    || !node.getNodeName().equals(ITEM_NODE)) {
                continue;
            }
            final Element timeNode = findChildElementByNodeName(node, TIME_NODE);
            if (timeNode == null || TextUtils.isEmpty(timeNode.getTextContent())) {
                continue;
            }
            try {
                result.add(parseTimeNode(timeNode.getTextContent()));
            } catch (final IllegalArgumentException ignored) {
            }
        }
        return result;
    }

    @Nullable
    private Element findChildElementByNodeName(
            @NonNull final Node parent,
            @NonNull final String nodeName) {
        for (Node child = parent.getFirstChild(); child != null; child = child.getNextSibling()) {
            if (child.getNodeType() != Node.ELEMENT_NODE) {
                continue;
            }
            if (nodeName.equals(child.getNodeName())) {
                return (Element) child;
            }
        }
        return null;
    }

    /**
     * 時間を表現する文字列をミリ秒に変換する。
     *
     * <p>フォーマット：0.000
     *
     * @param timeNode 時間文字列
     * @return ミリ秒
     */
    private int parseTimeNode(@NonNull final String timeNode) {
        try {
            return (int) (Float.parseFloat(timeNode) * TimeUnit.SECONDS.toMillis(1));
        } catch (final NumberFormatException e) {
            throw new IllegalArgumentException(e);
        }
    }
}

File: mobile/src/main/java/net/mm2d/android/upnp/cds/chapter/SonyFetcherFactory.kt
/*
 * Copyright (c) 2017 大前良介 (OHMAE Ryosuke)
 *
 * This software is released under the MIT License.
 * http://opensource.org/licenses/MIT
 */

package net.mm2d.android.upnp.cds.chapter

import io.reactivex.Single
import io.reactivex.SingleEmitter
import io.reactivex.schedulers.Schedulers
import net.mm2d.android.upnp.cds.CdsObject
import net.mm2d.upnp.HttpClient
import net.mm2d.upnp.util.XmlUtils
import net.mm2d.upnp.util.forEachElement
import org.w3c.dom.Element
import org.w3c.dom.Node
import org.xml.sax.SAXException
import java.io.IOException
import java.net.URL
import java.util.concurrent.TimeUnit
import javax.xml.parsers.ParserConfigurationException

/**
 * @author [大前良介 (OHMAE Ryosuke)](mailto:ryo@mm2d.net)
 */
internal class SonyFetcherFactory : FetcherFactory {
    override fun create(
        cdsObject: CdsObject,
    ): Single<List<Int>>? {
        val url = cdsObject.getValue(CHAPTER_INFO)
        if (url.isNullOrEmpty()) {
            return null
        }
        return Single.create { emitter: SingleEmitter<List<Int>> ->
            try {
                val xml = HttpClient(false).downloadString(URL(url))
                emitter.onSuccess(parseChapterInfo(xml))
            } catch (ignored: Exception) {
                emitter.onSuccess(emptyList())
            }
        }.subscribeOn(Schedulers.io())
    }

    @Throws(ParserConfigurationException::class, SAXException::class, IOException::class)
    private fun parseChapterInfo(
        xml: String,
    ): List<Int> {
        if (xml.isEmpty()) {
            return emptyList()
        }
        val root = XmlUtils.newDocument(false, xml).documentElement
        if (root == null || root.nodeName != ROOT_NODE) {
            return emptyList()
        }
        val content = root.findChildElementByNodeName(LIST_NODE) ?: return emptyList()
        val result = ArrayList<Int>()
        content.firstChild.forEachElement { element ->
            if (element.nodeName == ITEM_NODE) {
                element.findChildElementByNodeName(TIME_NODE)
                    ?.textContent
                    ?.parseTimeNode()
                    ?.let { result.add(it) }
            }
        }
        return result
    }

    private fun Node.findChildElementByNodeName(
        localName: String,
    ): Element? {
        firstChild?.forEachElement {
            if (localName == it.nodeName) {
                return it
            }
        }
        return null
    }

    /**
     * 時間を表現する文字列をミリ秒に変換する。
     *
     * フォーマット：0.000
     *
     * @receiver 時間文字列
     * @return ミリ秒
     */
    private fun String?.parseTimeNode(): Int? {
        if (this.isNullOrEmpty()) return null
        val time = this.toFloatOrNull() ?: return null
        return (time * ONE_SECOND).toInt()
    }

    companion object {
        private const val CHAPTER_INFO = "av:chapterInfo"
        private const val ROOT_NODE = "contentInfo"
        private const val LIST_NODE = "content_chapter_info"
        private const val ITEM_NODE = "chapter"
        private const val TIME_NODE = "chapter_point"
        private val ONE_SECOND = TimeUnit.SECONDS.toMillis(1)
    }
}
